### ISSUE TITLE
Only 2 different ads displayed

### ISSUE THREAD

**Issue by loki666** [clockbyte/admobadapter#60]
No matter what I set to setLimitOfAds(), only two different ads are displayed.

ie: list with 30 items,
mAdapterWrapper.setNoOfDataBetweenAds(8);
mAdapterWrapper.setLimitOfAds(3);

I do have 3 NativeExpressAd in the list, but first and third are the same object.
If i dismiss the first ad by clicking the little blue cross in the top right corner, the third ad is also dismissed.

I tried with setLimitOfAds(6) with bigger list, and it's the same result, only two different ads are displayed A - B - A - B - A ....

**Comment by kot331107** [comment]
@loki666 Sounds like correct behavior :) See https://github.com/clockbyte/admobadapter/wiki/Cookbook#in-case-you-want-to-use-a-multiple-unit-ids to publish different ads.

**Comment by loki666** [comment]
Either I wasn't clear enough, or I didn't understood how the AdapterWrapper is working...

I only have one ad unit id, and I'm asking for a maximum of 3 ads in my list.
What I'm observing, is that 1st and 3rd AdView show the same ads (ok It might happen), but when I dismiss the 1st AdView (by clicking the little blue cross) the 3rd ad is also dissmissed.

Which leads me to believe that the AdapterWrapper was reusing the same AdView instance for the 1st and 3rd ad.

So even, if I ask for 3 (or 4, or 10) ads, I actually get only 2 different AdView, reused in the list.

I though (maybe wrongly), that the Adapter would create 3 different AdView with 3 different AdRequest for the same Ad unit Id.

**Comment by loki666** [comment]
ok a bit more debugging.

It seems that there is only 2 different ads in my list (even if I request 3) because my list only require 2 ViewHolder for the Ads.

I would have expected of the Adapter, to indeed create only 2 ViewHolder, but still request 3 different ads, recycle the ViewHolder but still bind the 3 ads.

Is that the intended behavior of the Adapter?

**Comment by kot331107** [comment]
@loki666 You are quite right, Adapter would request N (almost) times given max ads count = N. But it's not a recycling issue or something. The content of ad block is filled by Admob. So if you want to publish N different ads and you want an expected behavior then please create N unit ids and pass its to admobadapter like it is shown  in wiki :)
Also...Do you use Express ads, right?

**Comment by loki666** [comment]
Yes I do use Express Ads,

following the code I don't think it's working like that...

the onBindViewHolder() does nothing if it's an AdExpress type. (where it should attach the NativeExpressAdView corresponding to the position in the ViewHolder's View)

only onCreateViewHolder() creates a Wrapper which hold the NativeExpressAdView.

In my case the RecyclerView needs only 2 ViewHolder for the Ads, but It depends on the setNoOfDataBetweenAds() and the screen size.

If I would have set a higher value (or on a smaller screen), the RecyclerView would only need 1 ViewHolder for the ad type,
resulting in only one NativeExpressAdView instance recycled in my list, no matter which value I set to setLimitOfAds().
setLimitOfAds affect the number of ads item inserted in the RecyclerView, not the number of NativeExpressAdView instances used in the list.

I also tried to use the constructor which use the ExpressAdPreset, but the result is exactly the same.
Only two ViewHolder needed, so only two NativeExpressAdView created recycled trough the list.

There should be as much NativeExpressAdView instances as setLimitOfAds() (less only if there is not enough items in the wrapped Adapter, but let's assume the list big enough), each NativeExpressAdView send its own AdRequest (via the Fetcher). The onBindViewHolder() should attach the correct NativeExpressAdView instance in the recycled ViewHolders.

**Comment by kot331107** [comment]
@loki666 The onBindViewHolder does nothing because admobadapter currently doesn't support ViewHolder pattern for express ads, but only for source (your data) items. So it works in the following way:
At 1st adapter wrapper would fetch and cache 2 ad blocks for future, see **AdmobExpressAdapterWrapper#init()** the last string `prefetchAds(AdmobFetcherExpress.PREFETCHED_ADS_SIZE);`
When you scrolling down your RecyclerView adapter wrapper calculates each next displaying item is of an Ad type. Then it gets the ad from **AdmobFetcherExpress#getAdForIndex(int adPos)**. If it was already stored for this position then Fetcher would return it back (express ads should refresh its' content themselves so we don't care about refetching it manually). Otherwise Fetcher would fetch one and cache it for this position. See **AdmobExpressAdapterWrapper#onCreateViewHolder(...)**
```java
NativeExpressAdViewHolder item = adFetcher.getAdForIndex(mCntAdCreated++);
if (item == null)
    item = prefetchAds(1);
```

You'd pay attention to Admob wouldn't provide you N different ads for the single unit id. Also I don't know why Admob returns cycling ad views for the single unit id in the list but I've noticed it too. Please try to make a simple test app with an Activity which contains 2 ad views using the same unit id (without recyclerview, caching, viewholders and other stuff like that). What will be a result then?

**Comment by loki666** [comment]
could we continue this on gitter.im ? it will be more convenient

**Comment by kot331107** [comment]
@loki666 sure

**Comment by kot331107** [comment]
https://github.com/googleads/googleads-mobile-android-examples/blob/master/advanced/NativeExpressRecyclerViewExample/app/src/main/java/com/google/android/gms/example/nativeexpressrecyclerviewexample/RecyclerViewAdapter.java the correct implementation of a binding (ViewHolder pattern for express native ads).

**Comment by kot331107** [comment]
@loki666 Hi, I've reimplemented the ads loading and presentation in admobadapter due to your request. Now admobadapter should handle the recycling of both the ListView and RecyclerView. You could update from the **Experimental** branch to get it. Soon it will be available in the Release branch with tag 1.4.0 and in the maven central too. Thanks a lot for your observations and investigations!
So closing it for now, but feel free to write if you'll find any issue.
